# Lab 2  Rong Xu(Cherie)
# Option 1: Fixed/ Random effect model

In [20]:
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy.linalg as la

In [5]:
data = pd.read_csv("trends-gss.csv")
#interested variable
sub = data[['year', 'age', 'sex', 'race','educ','paeduc','maeduc','speduc', 'region', 'marital']]

In [6]:
sub=sub.dropna()
sub.head()

year   age  sex  race  educ  paeduc  maeduc  speduc  region  marital
1  1972  70.0    1     1  10.0     8.0     8.0    12.0       3      1.0
2  1972  48.0    2     1  12.0     8.0     8.0    11.0       3      1.0
3  1972  27.0    2     1  17.0    16.0    12.0    20.0       3      1.0
4  1972  61.0    2     1  12.0     8.0     8.0    12.0       3      1.0
9  1972  30.0    2     2  12.0    10.0     7.0    11.0       7      1.0

In [7]:
# set index for panel data
sub = sub.set_index(['region','race'])
sub.head()

year   age  sex  educ  paeduc  maeduc  speduc  marital
region race                                                        
3      1     1972  70.0    1  10.0     8.0     8.0    12.0      1.0
       1     1972  48.0    2  12.0     8.0     8.0    11.0      1.0
       1     1972  27.0    2  17.0    16.0    12.0    20.0      1.0
       1     1972  61.0    2  12.0     8.0     8.0    12.0      1.0
7      2     1972  30.0    2  12.0    10.0     7.0    11.0      1.0

In [8]:
# check the count after remove na values - for example educ
sub['educ'].value_counts()

12.0    7073
16.0    3238
14.0    2310
13.0    1721
18.0     983
15.0     931
11.0     816
17.0     807
10.0     711
8.0      692
20.0     571
9.0      475
19.0     365
7.0      202
6.0      188
5.0       82
4.0       65
3.0       51
2.0       33
0.0       31
1.0        9
Name: educ, dtype: int64

## (a) Run an OLS regression, including at least one independent variable and a time variable (as dummies).  Explain how you think your independent variable relates to your dependent variable.  Interpret your results.  Did you find what you expected to find? 

In [9]:
#runs an OLS regression, including at least one independent variable and a time variable (as dummies) 

lm1 = smf.ols('educ ~ paeduc + year', data=sub).fit()
lm1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.258
Model:                            OLS   Adj. R-squared:                  0.258
Method:                 Least Squares   F-statistic:                     3717.
Date:                Sat, 11 Nov 2023   Prob (F-statistic):               0.00
Time:                        22:55:06   Log-Likelihood:                -50771.
No. Observations:               21354   AIC:                         1.015e+05
Df Residuals:                   21351   BIC:                         1.016e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -53.4734      3.075    -17.389      0.000     -59.501     -47.446
paeduc         0.3297      0.004     76.778      0.000       0.321       0.338
year           0.0318      0.002     20.539      0.000       0.029       0.035
==============================================================================
Omnibus:                      498.960   Durbin-Watson:                   1.838
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1155.096
Skew:                           0.040   Prob(JB):                    1.49e-251
Kurtosis:                       4.137   Cond. No.                     3.43e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.43e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### for this regression model, I chose paedu as independent variable, year as dummies variable educ as dependent variable that I want to see how father's education will affect their kids' education attainment. I assume, people will have higher education attainment with year passes, also their father's degree will positive affect them to gain higher education. In order to verify my assumption, we can look at the output from the linear regression model. 
### R-square = 0.258, it presents that 25.8% variance of educ can be explained by paedu and year. Constant the year, the father gain one addtional unit of education, their kid's education will raise up by 0.3297 unit, at a statistcally significant level with p-value =0. Constant the paeduc, the respondent's education goes up with year passes by 0.0318 unit, at a statisitcally significant. Therefore, we can say paeduc and year have positive correlation with educ and it's less likely to happen it by chance with this low se and high t-stats value, and we can reject null hypothesis. 


# (b) Then run a fixed effect model version of that OLS model.  Interpret your results.  Did you find what you expected to find?  Why?  Why not? 

In [10]:
#  fixed effect model version of that OLS model. 
lm2 = smf.ols('educ ~ paeduc + C(year)', data=sub).fit()
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.260
Model:                            OLS   Adj. R-squared:                  0.259
Method:                 Least Squares   F-statistic:                     258.2
Date:                Sat, 11 Nov 2023   Prob (F-statistic):               0.00
Time:                        22:55:12   Log-Likelihood:                -50747.
No. Observations:               21354   AIC:                         1.016e+05
Df Residuals:                   21324   BIC:                         1.018e+05
Df Model:                          29                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           9.1807      0.105     87.766      0.000       8.976       9.386
C(year)[T.1973]     0.0843      0.136      0.619      0.536      -0.183       0.351
C(year)[T.1974]     0.2776      0.138      2.018      0.044       0.008       0.547
C(year)[T.1975]     0.0156      0.137      0.114      0.909      -0.252       0.283
C(year)[T.1976]     0.2727      0.140      1.946      0.052      -0.002       0.547
C(year)[T.1977]     0.1864      0.142      1.315      0.188      -0.091       0.464
C(year)[T.1978]     0.3073      0.141      2.178      0.029       0.031       0.584
C(year)[T.1980]     0.2857      0.144      1.990      0.047       0.004       0.567
C(year)[T.1982]     0.3703      0.141      2.617      0.009       0.093       0.648
C(year)[T.1983]     0.5027      0.140      3.590      0.000       0.228       0.777
C(year)[T.1984]     0.6593      0.145      4.536      0.000       0.374       0.944
C(year)[T.1985]     0.4926      0.143      3.453      0.001       0.213       0.772
C(year)[T.1986]     0.6598      0.145      4.539      0.000       0.375       0.945
C(year)[T.1987]     0.6412      0.141      4.548      0.000       0.365       0.918
C(year)[T.1988]     0.6700      0.147      4.563      0.000       0.382       0.958
C(year)[T.1989]     0.7788      0.143      5.448      0.000       0.499       1.059
C(year)[T.1990]     0.7689      0.151      5.085      0.000       0.473       1.065
C(year)[T.1991]     0.9493      0.149      6.357      0.000       0.657       1.242
C(year)[T.1993]     0.9771      0.142      6.874      0.000       0.698       1.256
C(year)[T.1994]     0.9633      0.125      7.688      0.000       0.718       1.209
C(year)[T.1996]     1.1109      0.128      8.651      0.000       0.859       1.363
C(year)[T.1998]     0.9806      0.129      7.579      0.000       0.727       1.234
C(year)[T.2000]     1.0341      0.131      7.894      0.000       0.777       1.291
C(year)[T.2002]     1.1306      0.130      8.706      0.000       0.876       1.385
C(year)[T.2004]     1.2437      0.126      9.895      0.000       0.997       1.490
C(year)[T.2006]     0.9989      0.126      7.928      0.000       0.752       1.246
C(year)[T.2008]     1.2177      0.138      8.812      0.000       0.947       1.489
C(year)[T.2010]     1.1610      0.141      8.205      0.000       0.884       1.438
C(year)[T.2012]     1.2405      0.140      8.849      0.000       0.966       1.515
paeduc              0.3284      0.004     76.320      0.000       0.320       0.337
==============================================================================
Omnibus:                      492.562   Durbin-Watson:                   1.842
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1130.402
Skew:                           0.044   Prob(JB):                    3.44e

### In fixed model, the R-square = 0.26, which increased compared with lm1 (R-sqaure=0.258), which improve the posibility of how more likely the variance of educ can be explained by paeduc and year. Look at year, for year 1975, coeff=0.0156, we can say there is a big drop on 1975 for people who will higher education, at a statistcally insignificant level of p-value=0.909.  Holding all other variables constant. In this case, it suggests that, compared to the reference year(year 1973 at the begining), the average level of education (educ) is expected to increase by 0.3703 units in the year 1982. After year 1982, the coeff had have raising with statistically significant level of p-value lower than 0.01. Constant years, with father's education level rasie up by one unit, the kid's education will increase by 0.3284 unit at statistically significant level on averge. Compared with lm1 paeduc(coeff)=0.3297, it decreased by 0.013 unit in the fixed effect model. Overall, this model agreed that the paedu and year have positive effect on educ, more importantly, we could focus on the observation after year 1982 as these p-value present they are statistically significant and could tell us how 'year' will affect on educ. 

### (c) Then include an additional predictor in your fixed effects model that you think might account for the initial relationship you found between your X and your Y.  What effect does that new independent variable have in your new regression?

In [14]:
# add maeduc in
lm3 = smf.ols('educ ~ paeduc + maeduc + C(year)', data=sub).fit()
lm3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.298
Model:                            OLS   Adj. R-squared:                  0.297
Method:                 Least Squares   F-statistic:                     301.9
Date:                Sat, 11 Nov 2023   Prob (F-statistic):               0.00
Time:                        23:01:11   Log-Likelihood:                -50181.
No. Observations:               21354   AIC:                         1.004e+05
Df Residuals:                   21323   BIC:                         1.007e+05
Df Model:                          30                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           8.2810      0.105     78.691      0.000       8.075       8.487
C(year)[T.1973]     0.0319      0.133      0.240      0.810      -0.228       0.292
C(year)[T.1974]     0.1710      0.134      1.276      0.202      -0.092       0.434
C(year)[T.1975]    -0.0913      0.133     -0.686      0.493      -0.352       0.169
C(year)[T.1976]     0.2439      0.137      1.787      0.074      -0.024       0.511
C(year)[T.1977]     0.0838      0.138      0.607      0.544      -0.187       0.354
C(year)[T.1978]     0.1561      0.137      1.135      0.256      -0.113       0.426
C(year)[T.1980]     0.1827      0.140      1.307      0.191      -0.091       0.457
C(year)[T.1982]     0.2375      0.138      1.723      0.085      -0.033       0.508
C(year)[T.1983]     0.3999      0.136      2.932      0.003       0.133       0.667
C(year)[T.1984]     0.4881      0.142      3.446      0.001       0.211       0.766
C(year)[T.1985]     0.3623      0.139      2.607      0.009       0.090       0.635
C(year)[T.1986]     0.4815      0.142      3.400      0.001       0.204       0.759
C(year)[T.1987]     0.4311      0.137      3.137      0.002       0.162       0.701
C(year)[T.1988]     0.4605      0.143      3.217      0.001       0.180       0.741
C(year)[T.1989]     0.6158      0.139      4.421      0.000       0.343       0.889
C(year)[T.1990]     0.5978      0.147      4.057      0.000       0.309       0.887
C(year)[T.1991]     0.7873      0.146      5.411      0.000       0.502       1.072
C(year)[T.1993]     0.7937      0.139      5.730      0.000       0.522       1.065
C(year)[T.1994]     0.7796      0.122      6.384      0.000       0.540       1.019
C(year)[T.1996]     0.8469      0.125      6.759      0.000       0.601       1.093
C(year)[T.1998]     0.7710      0.126      6.111      0.000       0.524       1.018
C(year)[T.2000]     0.7999      0.128      6.261      0.000       0.549       1.050
C(year)[T.2002]     0.8874      0.127      7.006      0.000       0.639       1.136
C(year)[T.2004]     1.0090      0.123      8.230      0.000       0.769       1.249
C(year)[T.2006]     0.8401      0.123      6.842      0.000       0.599       1.081
C(year)[T.2008]     1.0033      0.135      7.448      0.000       0.739       1.267
C(year)[T.2010]     0.9492      0.138      6.881      0.000       0.679       1.220
C(year)[T.2012]     1.0080      0.137      7.375      0.000       0.740       1.276
paeduc              0.2047      0.006     36.918      0.000       0.194       0.216
maeduc              0.2214      0.006     34.078      0.000       0.209       0.234
==============================================================================
Omnibus:                      528.896   Durbin-Watson:                   1.849
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            

#### I chose to add in maeduc this predictor variable that to see how parental education will likely affect their kid's education attainment. In this new model, R-square = 0.298, which bigger than lm2 (R-sqaure=0.26), lm1 (R-sqaure=0.258), and it increase the possibility of how the variance of educ can explained by paeduc, maeduc, and year. While, it could also becasue of with new independent variable add in, the R-square will be likely improved anyway. paeduc(B)=0.2047, which lower than lm1(0.3927) and lm2(0.3284), but it indicated that on average, one unit increase on father education, kids' education will increase by 0.2047 at statistically significant. For maeduc, one unit increase of it, kid's education will positively effect by increasing 0.2214 unit at statistcally siginificant on average. In year 1983, coeff (year 1983) = 0.3999, it present that holding all other variables constant, it suggests that, compared to the reference year(year 1973 at the begining), the average level of education (educ) is expected to increase by 0.3999 units in the year 1983 at statistcally significant with p-value = 0.003. After 1983, all p-value in year are lower than 0.05 which tells they are statistically significant and positively affect on educ. Compared with lm2, it change the breakdown on 1982 rather than 1983, so maeduc helped us to more accurate predict our assumption. 

# Then run a random effects model equivalent to your fixed effects model in step (b).  Interpret the results.

In [15]:
from linearmodels.panel import RandomEffects

# run the random effects model equivalent from lm2
lm4 = RandomEffects.from_formula('educ ~ paeduc + C(year)', data=sub).fit()
lm4.summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                        RandomEffects Estimation Summary                        
================================================================================
Dep. Variable:                   educ   R-squared:                        0.2599
Estimator:              RandomEffects   R-squared (Between):              0.7331
No. Observations:               21354   R-squared (Within):               0.2533
Date:                Sat, Nov 11 2023   R-squared (Overall):              0.2599
Time:                        23:03:39   Log-likelihood                -5.075e+04
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      258.19
Entities:                           9   P-value                           0.0000
Avg Obs:                       2372.7   Distribution:                F(29,21324)
Min Obs:                       1042.0                                           
Max Obs:                       4228.0   F-statistic (robust):          1.901e+04
                                        P-value                           0.0000
Time periods:                       3   Distribution:                F(29,21324)
Avg Obs:                       7118.0                                           
Min Obs:                       881.00                                           
Max Obs:                    1.903e+04                                           
                                                                                
                                Parameter Estimates                                
===================================================================================
                 Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
-----------------------------------------------------------------------------------
paeduc              0.3284     0.0043     76.320     0.0000      0.3199      0.3368
C(year)[T.1972]     9.1807     0.1046     87.766     0.0000      8.9756      9.3857
C(year)[T.1973]     9.2650     0.1024     90.481     0.0000      9.0643      9.4657
C(year)[T.1974]     9.4583     0.1043     90.673     0.0000      9.2538      9.6628
C(year)[T.1975]     9.1963     0.1029     89.360     0.0000      8.9945      9.3980
C(year)[T.1976]     9.4534     0.1078     87.669     0.0000      9.2420      9.6648
C(year)[T.1977]     9.3671     0.1097     85.362     0.0000      9.1520      9.5822
C(year)[T.1978]     9.4880     0.1097     86.522     0.0000      9.2730      9.7029
C(year)[T.1980]     9.4664     0.1129     83.823     0.0000      9.2450      9.6877
C(year)[T.1982]     9.5509     0.1104     86.544     0.0000      9.3346      9.7672
C(year)[T.1983]     9.6834     0.1092     88.693     0.0000      9.4694      9.8974
C(year)[T.1984]     9.8399     0.1157     85.056     0.0000      9.6132      10.067
C(year)[T.1985]     9.6733     0.1128     85.793     0.0000      9.4523      9.8943
C(year)[T.1986]     9.8405     0.1159     84.935     0.0000      9.6134      10.068
C(year)[T.1987]     9.8219     0.1103     89.039     0.0000      9.6057      10.038
C(year)[T.1988]     9.8506     0.1178     83.636     0.0000      9.6198      10.081
C(year)[T.1989]     9.9594     0.1126     88.452     0.0000      9.7387      10.180
C(year)[T.1990]     9.9496     0.1241     80.171     0.0000      9.7063      10.193
C(year)[T.1991]     10.130     0.1219     83.131     0.0000      9.8911      10.369
C(year)[T.1993]     10.158     0.1131     89.847     0.0000      9.9361      10.379
C(year)[T.1994]     10.144     0.0914     111.04     0.0000      9.9649      10.323
C(year)[T.1996]     10.292     0.0960     107.17     0.0000      10.103      10.480
C(year)[T.1998]     10.161     0.0972     104.55     0.0000      9.9708      10.352
C(year)[T.2000]     10.215     0.0993     102.82     0.0000      10.020      10.410
C(year)[T.2002]     10.311     0.0981     105.11     0.

### In this model, R-square = 0.2599, which is lower than lm3(0.298), it tells that 25.99% of variance in educ can be explained by paeduc and year. parameter paeduc = 0.3284, it is same as the coefficient of paeduc in lm2 from previous in fixed effect model. It has a very low p-value (0.0000), suggesting a statistically significant positive relationship between father education and the education level of the individual.
### For years, all of these coefficients are highly statistically significant (p-values are 0.0000). For year 1975, the average education level is estimated to be 0.1963 units higher than the reference year(year 1972), constant the paeduc. It suggest an increase in education level for the year 1975 relative to the refernce year and this observed change is unlikely to be due to random chance. After year 1975, the parameter generally presents a trend upward state that with year goes up, the individual education level will raise faster compared with the reference year 1972. 
### R-square (between) = 0.7331, it tells that the proportion of total variability in educ attributed to differences between paedu that grouped by year. R-square(within) = 0,2533, it tells that the proportion of total variability in educ attributed to differences within paeduc over time. It indicate that a substantial portion of the total variability in educ is attributed to difference between years and within years over time. 
### Overall, the random effects model suggests that both father's education (paeduc) and the year of observation (C(year)) have statistically significant effects on the education level (educ). 

## Run a Hausman test to compare your fixed effects and your random effects models.  What do you conclude? 

In [19]:
from linearmodels.panel import PanelOLS, RandomEffects

y = sub['educ']
x = sub[['paeduc', 'year']]

# Fixed Effects Model
lm2 = PanelOLS(y, x, entity_effects=True, time_effects=True).fit()

# Random Effects Model
lm4 = RandomEffects(y, x).fit()

# Wald test for exogeneity in Random Effects Model
wald_test = lm4.wald_test(formula='paeduc = 0, year = 0')

# Print Wald test results
print(wald_test)

Linear Equality Hypothesis Test
H0: Linear equality constraint is valid
Statistic: 210793.0193
P-value: 0.0000
Distributed: chi2(2)


### in this case, the p-value from the test is very small, it suggests that the fixed effects model is preferred.  Also, with the pvalue = 0, we can reject the null hypothesis, indicating that the coefficients of the specified variables are not jointly equal to zero. 
### compared with lm2(fixed effect model) and lm4(random effect model), the R-square is higher on lm2 with the constant value on paedu of coeff. Also, lm2 has better clear on dummy variable year that it tells us specific time period we could focus and observe the effect between year and educ, while lm4 didn't work well on it as it shows entire time period are statistically significant. 
### with the help of hausman test, I would like to conclude that fixed effect model works better in this case. 